runtime/src/accounts_db.rs

```diff
@@ -3660,79 +3656,205 @@ impl AccountsDB {
             .collect()
     }
 
-    fn remove_zero_balance_accounts(
-        account_maps: DashMap<Pubkey, CalculateHashIntermediate>,
-    ) -> Vec<(Pubkey, Hash, u64)> {
-        type ShardType = dashmap::lock::RwLock<
-            std::collections::HashMap<
-                solana_sdk::pubkey::Pubkey,
-                dashmap::SharedValue<CalculateHashIntermediate>,
-            >,
-        >;
-        let shards: &[ShardType] = account_maps.shards();
-
-        let hashes: Vec<_> = shards
-            .par_iter()
-            .map(|x| {
-                let a: dashmap::lock::RwLockReadGuard<HashMap<_, _>> = x.read();
-                let res: Vec<_> = a
-                    .iter()
-                    .filter_map(|inp| {
-                        let (pubkey, sv) = inp;
-                        let item = sv.get();
-                        if item.raw_lamports != 0 {
-                            Some((*pubkey, item.hash, item.lamports))
-                        } else {
-                            None
-                        }
-                    })
-                    .collect();
-                res
+    fn flatten_hash_intermediate(
+        data_sections_by_pubkey: Vec<Vec<CalculateHashIntermediate>>,
+    ) -> (Vec<CalculateHashIntermediate>, Measure, usize) {
+        let mut flatten_time = Measure::start("flatten");
+        let result: Vec<_> = data_sections_by_pubkey.into_iter().flatten().collect();
+        let raw_len = result.len();
+        flatten_time.stop();
+        (result, flatten_time, raw_len)
+    }
+
+    fn compare_two_hash_entries(
+        a: &CalculateHashIntermediate,
+        b: &CalculateHashIntermediate,
+    ) -> std::cmp::Ordering {
+        match a.pubkey.cmp(&b.pubkey) {
+            std::cmp::Ordering::Equal => match a.slot.cmp(&b.slot).reverse() {
+                std::cmp::Ordering::Equal => a.version.cmp(&b.version).reverse(),
+                other => other,
+            },
+            other => other,
+        }
+    }
+
+    fn sort_hash_intermediate(
+        mut data_by_pubkey: Vec<CalculateHashIntermediate>,
+    ) -> (Vec<CalculateHashIntermediate>, Measure) {
+        // sort each PUBKEY_DIVISION vec
+        let mut sort_time = Measure::start("sort");
+        data_by_pubkey.par_sort_unstable_by(Self::compare_two_hash_entries);
+        sort_time.stop();
+        (data_by_pubkey, sort_time)
+    }
+
+    fn de_dup_and_eliminate_zeros(
+        sorted_data_by_pubkey: Vec<CalculateHashIntermediate>,
+    ) -> (Vec<Vec<Hash>>, Measure, u64) {
+        // 1. eliminate zero lamport accounts
+        // 2. pick the highest slot or (slot = and highest version) of each pubkey
+        // 3. produce this output:
+        // vec: PUBKEY_BINS_FOR_CALCULATING_HASHES in pubkey order
+        //   vec: sorted sections from parallelism, in pubkey order
+        //     vec: individual hashes in pubkey order
+        let mut zeros = Measure::start("eliminate zeros");
+        const CHUNKS: usize = 10;
```

Comment:
Should this be based on number of cpus?

Reply:
This is also another constant number with not enough intuitive understanding yet. Empirically, 10 works better than 1. Once it is A number, my thinking is the algorithm can be tuned to 15M accounts, 80M, 100M, 1B, etc. My process is: Measure. See something expensive. Think about how to make it faster. Try something. Repeat. I agree that the tuning of any of this is not well done yet. I was making some assumptions like: machines have lots of cpus. There is lots of data. cpus and data likely will grow in the future. So, find places where we can split up work and see what effect that has and what side effects it introduces. And, then, the levers are there ready to pull as we hit the next hotspot. I'm open to feedback. 

But, also, this is the simpler changes I pulled out of the changes that were twice as fast as this. My thinking in this PR is these ideas seem sufficiently sound to build off and they have enough performance bang to be a good tradeoff for the moment while we keep getting better as time marches forward.

Again, I don't mean to sound defensive. Text is hard. I am trying to discuss/explore the tradeoffs and get us on the same page.